File: LeetCode/python/lc_0209.py
"""
 * @file     lc_0209.py
 * @brief    https://leetcode-cn.com/problems/minimum-size-subarray-sum/
 * @author   YongDu
 * @date     2021-09-09
"""


class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        min_len = sys.maxsize
        left = 0
        sum = 0
        for i in range(len(nums)):
            sum += nums[i]
            while(sum >= target):
                min_len = min(i - left + 1, min_len)
                sum -= nums[left]
                left += 1
            i += 1
        return min_len if min_len != sys.maxsize else 0

File: LeetCode/cpp/lc_0209.cpp
/**
 * @file     lc_0209.cpp
 * @brief    https://leetcode-cn.com/problems/minimum-size-subarray-sum/
 * @author   YongDu
 * @date     2021-09-07
 */
class Solution {
public:
  int minSubArrayLen(int target, vector<int> &nums) {
    int minLen = INT_MAX;
    int left = 0;
    int sum = 0;
    for (int right = 0; right < nums.size(); ++right) {
      sum += nums[right];
      // 缩小滑动窗口
      while (sum >= target) {
        minLen = std::min(right - left + 1, minLen);
        sum -= nums[left++];
      }
    }
    return minLen != INT_MAX ? minLen : 0;
  }
};
